trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/NwaCegarLoop.java

```diff
@@ -0,0 +1,565 @@
+import de.uni_freiburg.informatik.ultimate.plugins.generator.traceabstraction.preferences.TraceAbstractionPreferenceInitializer.CounterexampleSearchStrategy;
+import de.uni_freiburg.informatik.ultimate.plugins.generator.traceabstraction.preferences.TraceAbstractionPreferenceInitializer.Minimization;
+import de.uni_freiburg.informatik.ultimate.plugins.generator.traceabstraction.preferences.TraceAbstractionPreferenceInitializer.RelevanceAnalysisMode;
+import de.uni_freiburg.informatik.ultimate.util.HistogramOfIterable;
+
+/**
+ * Subclass of BasicCegarLoop for safety checking based on nested-word automata.
+ *
+ * @author Matthias Heizmann (heizmann@informatik.uni-freiburg.de)
+ * @author Christian Schilling (schillic@informatik.uni-freiburg.de)
+ * @author Dominik Klumpp (klumpp@informatik.uni-freiburg.de)
+ */
+public class NwaCegarLoop<L extends IIcfgTransition<?>> extends BasicCegarLoop<L, INestedWordAutomaton<L, IPredicate>> {
+
+	protected static final int MINIMIZE_EVERY_KTH_ITERATION = 10;
+	protected static final boolean REMOVE_DEAD_ENDS = true;
+	protected static final int MINIMIZATION_TIMEOUT = 1_000;
+
+	/**
+	 * If the trace histogram max is larger than this number we try to find a danger invariant. Only used for debugging.
+	 */
+	private static final int DEBUG_DANGER_INVARIANTS_THRESHOLD = Integer.MAX_VALUE;
+
+	protected final Collection<INwaOutgoingLetterAndTransitionProvider<L, IPredicate>> mStoredRawInterpolantAutomata;
+
+	private final SearchStrategy mSearchStrategy;
+	private final ErrorGeneralizationEngine<L> mErrorGeneralizationEngine;
+
+	private final boolean mUseHeuristicEmptinessCheck;
+	private final ScoringMethod mScoringMethod;
+	private final AStarHeuristic mAStarHeuristic;
+	private final Integer mAStarRandomHeuristicSeed;
+
+	public NwaCegarLoop(final DebugIdentifier name, final INestedWordAutomaton<L, IPredicate> initialAbstraction,
+			final IIcfg<?> rootNode, final CfgSmtToolkit csToolkit, final PredicateFactory predicateFactory,
+			final TAPreferences taPrefs, final Set<? extends IcfgLocation> errorLocs,
+			final InterpolationTechnique interpolation, final boolean computeHoareAnnotation,
+			final Set<IcfgLocation> hoareAnnotationLocs, final IUltimateServiceProvider services,
+			final Class<L> transitionClazz, final PredicateFactoryRefinement stateFactoryForRefinement) {
+		super(name, initialAbstraction, rootNode, csToolkit, predicateFactory, taPrefs, errorLocs, interpolation,
+				computeHoareAnnotation, hoareAnnotationLocs, services, transitionClazz, stateFactoryForRefinement);
+
+		mErrorGeneralizationEngine = new ErrorGeneralizationEngine<>(services);
+
+		final IPreferenceProvider prefs = getServices().getPreferenceProvider(Activator.PLUGIN_ID);
+		mSearchStrategy = getSearchStrategy(prefs);
+		mStoredRawInterpolantAutomata = checkStoreCounterExamples(mPref) ? new ArrayList<>() : null;
+
+		// Heuristic Emptiness Check
+		mUseHeuristicEmptinessCheck = taPrefs.useHeuristicEmptinessCheck();
```

Comment:
The mix of different preference instances (mPrefs, taPrefs, getPreferenceProvider) and concepts (always get it from the instance, save it in the class, ...) is confusing. We should settle on one method with Matthias.

Reply:
Agreed.